From 2b4c3e73358bd18a5a89e04e23eb0effe9809fb3 Mon Sep 17 00:00:00 2001
From: mkundaragi <mkundaragi@75f.io>
Date: Wed, 15 Jun 2022 07:34:17 +0000
Subject: [PATCH 12/50] Merged PR 4141: VOC and pm2.5 configuration for HS &
 minor UI changes

Added new Configuration for VOC and Pm 2.5 sensor for alert
At device layer we are sending an highest value of VOC and PM 2.5  i.e 1000

Related work items: #11337, #11581
---
 .../hyperstat/cpu/HyperstatCpuFragment.kt     |  68 +++++-
 .../hyperstat/cpu/HyperstatCpuViewModel.kt    | 102 ++++++--
 .../layout/fragment_hyperstat_cpu_config.xml  |  95 ++++++++
 app/src/main/res/values/strings.xml           | 152 ++++++------
 app/src/main/res/values/styles.xml            |  23 +-
 .../hyperstat/HyperStatMessageGenerator.java  |  27 ++-
 .../hyperstat/common/HSReconfigureUtil.kt     |   2 +-
 .../cpu/HyperStatCpuConfiguration.kt          |   5 +
 .../hyperstat/cpu/HyperStatCpuEquip.kt        | 225 ++++++++++++------
 .../hyperstat/cpu/HyperStatPointsUtil.kt      |  60 ++++-
 .../a75f/io/logic/util/MigrationUtil.java     |  54 ++++-
 .../a75f/io/logic/util/PreferenceUtil.java    |  18 +-
 12 files changed, 644 insertions(+), 187 deletions(-)

diff --git a/app/src/main/java/a75f/io/renatus/hyperstat/cpu/HyperstatCpuFragment.kt b/app/src/main/java/a75f/io/renatus/hyperstat/cpu/HyperstatCpuFragment.kt
index b6288d913..c095d5fdf 100644
--- a/app/src/main/java/a75f/io/renatus/hyperstat/cpu/HyperstatCpuFragment.kt
+++ b/app/src/main/java/a75f/io/renatus/hyperstat/cpu/HyperstatCpuFragment.kt
@@ -15,7 +15,6 @@ import a75f.io.renatus.BASE.BaseDialogFragment
 import a75f.io.renatus.BASE.FragmentCommonBundleArgs
 import a75f.io.renatus.FloorPlanFragment
 import a75f.io.renatus.R
-import a75f.io.renatus.hyperstat.*
 import a75f.io.renatus.util.ProgressDialogUtils
 import a75f.io.renatus.util.RxjavaUtil
 import a75f.io.renatus.util.extension.showErrorDialog
@@ -61,10 +60,10 @@ class HyperStatCpuFragment : BaseDialogFragment() {
     lateinit var autoAwaySwitch: ToggleButton
 
     // 6 rows, 1 for each relay
-    lateinit var relayUIs: List<RelayWidgets>
+    private lateinit var relayUIs: List<RelayWidgets>
 
     // 3 rows, 1 for each analog out, plus damper voltage selectors
-    lateinit var analogOutUIs: List<AnalogOutWidgets>
+    private lateinit var analogOutUIs: List<AnalogOutWidgets>
 
     lateinit var airflowSensorSwitch: ToggleButton
     lateinit var doorWindowSensorSwitch: ToggleButton
@@ -74,9 +73,15 @@ class HyperStatCpuFragment : BaseDialogFragment() {
 
     lateinit var setButton: Button
     lateinit var zoneCO2Layout: View
-    lateinit var zoneCO2DamperOpeningRate: Spinner
-    lateinit var zoneCO2Threshold: Spinner
-    lateinit var zoneCO2Target: Spinner
+    private lateinit var zoneCO2DamperOpeningRate: Spinner
+    private lateinit var zoneCO2Threshold: Spinner
+    private lateinit var zoneCO2Target: Spinner
+
+    lateinit var zoneVOCThreshold: Spinner
+    lateinit var zoneVOCTarget: Spinner
+    lateinit var zonePMThreshold: Spinner
+    lateinit var zonePMTarget: Spinner
+
 
     /**
      * Test Signal Buttons
@@ -224,6 +229,11 @@ class HyperStatCpuFragment : BaseDialogFragment() {
             zoneCO2Threshold = findViewById(R.id.zoneCO2ThresholdSpinner)
             zoneCO2Target = findViewById(R.id.zoneCO2TargetSpinner)
 
+            zoneVOCThreshold = findViewById(R.id.zoneVocThresholdSpinner)
+            zoneVOCTarget = findViewById(R.id.zoneVocTargetSpinner)
+            zonePMThreshold = findViewById(R.id.zonepmThresholdSpinner)
+            zonePMTarget = findViewById(R.id.zonepmTargetSpinner)
+
             setButton = findViewById(R.id.setButton)
 
             /**
@@ -295,10 +305,36 @@ class HyperStatCpuFragment : BaseDialogFragment() {
             requireContext(), R.layout.larger_spinner_item,
             co2OpeningValues
         )
+
+        val vocAdapter: ArrayAdapter<*> = ArrayAdapter<String?>(
+            requireContext(), R.layout.larger_spinner_item,
+            vocValues()
+        )
+        val pmAdapter: ArrayAdapter<*> = ArrayAdapter<String?>(
+            requireContext(), R.layout.larger_spinner_item,
+            pmValues()
+        )
+
+
+
         zoneCO2DamperOpeningRate.adapter = co2OpeningAdapter
         zoneCO2Threshold.adapter = co2Adapter
         zoneCO2Target.adapter = co2Adapter
 
+        zoneVOCThreshold.adapter = vocAdapter
+        zoneVOCTarget.adapter = vocAdapter
+
+        zonePMThreshold.adapter = pmAdapter
+        zonePMTarget.adapter = pmAdapter
+
+
+        // set default values
+        zoneVOCThreshold.setSelection(zoneVOCThreshold.adapter.count -1)
+        zoneVOCTarget.setSelection(zoneVOCTarget.adapter.count -1)
+
+        zonePMThreshold.setSelection(zonePMThreshold.adapter.count -1)
+        zonePMTarget.setSelection(zonePMTarget.adapter.count -1)
+
 
         analogOutUIs.forEach {
 
@@ -409,10 +445,19 @@ class HyperStatCpuFragment : BaseDialogFragment() {
         zoneCO2Threshold.setOnItemSelected { position -> viewModel.zoneCO2ThresholdSelect(position) }
         zoneCO2Target.setOnItemSelected { position -> viewModel.zoneCO2TargetSelect(position) }
 
+        zoneVOCThreshold.setOnItemSelected { position ->viewModel.zoneVOCThresholdSelect(position)  }
+        zoneVOCTarget.setOnItemSelected { position ->viewModel.zoneVOCTargetSelect(position)  }
+        zonePMThreshold.setOnItemSelected { position ->viewModel.zonePmThresholdSelect(position)  }
+        zonePMTarget.setOnItemSelected { position ->viewModel.zonePmTargetSelect(position)  }
+
 
         // On Click save the CPU configuration
         setButton.setOnClickListener {
             setButton.isEnabled = false
+            Log.i("DEV_DEBUG", "zoneVOCThreshold: ${zoneVOCThreshold.selectedItem}")
+            Log.i("DEV_DEBUG", "zoneVOCTarget: ${zoneVOCTarget.selectedItem}")
+            Log.i("DEV_DEBUG", "zonePMThreshold: ${zonePMThreshold.selectedItem}")
+            Log.i("DEV_DEBUG", "zonePMTarget: ${zonePMTarget.selectedItem}")
             configurationDisposable.add(RxjavaUtil.executeBackgroundTaskWithDisposable(
                 {
                     ProgressDialogUtils.showProgressDialog(activity, "Saving Hyperstat CPU Configuration")
@@ -487,7 +532,7 @@ class HyperStatCpuFragment : BaseDialogFragment() {
                 analogOutAtFanMedium.setSelection(analogOutState.perAtFanMediumPosition)
                 analogOutAtFanHigh.setSelection(analogOutState.perAtFanHighPosition)
 
-                if (analogOutState.association.ordinal == CpuAnalogOutAssociation.DCV_DAMPER.ordinal) {
+                if (analogOutState.enabled && analogOutState.association.ordinal == CpuAnalogOutAssociation.DCV_DAMPER.ordinal) {
                     isDampSelected = true
                 }
             }
@@ -503,10 +548,17 @@ class HyperStatCpuFragment : BaseDialogFragment() {
                 selector.setSelection(analogInState.association.ordinal)
             }
         }
+
         zoneCO2Layout.visibility = if (isDampSelected) View.VISIBLE else View.GONE
         zoneCO2DamperOpeningRate.setSelection(viewState.zoneCO2DamperOpeningRatePos)
         zoneCO2Threshold.setSelection(viewState.zoneCO2ThresholdPos)
         zoneCO2Target.setSelection(viewState.zoneCO2TargetPos)
+
+        zoneVOCThreshold.setSelection(viewState.zoneVocThresholdPos)
+        zoneVOCTarget.setSelection(viewState.zoneVocTargetPos)
+
+        zonePMThreshold.setSelection(viewState.zonePm2p5ThresholdPos)
+        zonePMTarget.setSelection(viewState.zonePm2p5TargetPos)
     }
 
 
@@ -516,7 +568,7 @@ class HyperStatCpuFragment : BaseDialogFragment() {
         disposables.dispose()
         configurationDisposable.dispose()
         if (Globals.getInstance().isTestMode) {
-            Globals.getInstance().isTestMode = false;
+            Globals.getInstance().isTestMode = false
         }
     }
 
diff --git a/app/src/main/java/a75f/io/renatus/hyperstat/cpu/HyperstatCpuViewModel.kt b/app/src/main/java/a75f/io/renatus/hyperstat/cpu/HyperstatCpuViewModel.kt
index bd5ee28aa..5718d135c 100644
--- a/app/src/main/java/a75f/io/renatus/hyperstat/cpu/HyperstatCpuViewModel.kt
+++ b/app/src/main/java/a75f/io/renatus/hyperstat/cpu/HyperstatCpuViewModel.kt
@@ -1,6 +1,5 @@
 package a75f.io.renatus.hyperstat.cpu
 
-import a75f.io.logic.Globals
 import a75f.io.logic.L
 import a75f.io.logic.bo.building.NodeType
 import a75f.io.logic.bo.building.Output
@@ -15,7 +14,6 @@ import a75f.io.logic.bo.building.hyperstat.cpu.CpuAnalogOutAssociation.*
 import a75f.io.logic.bo.building.hyperstat.cpu.CpuRelayAssociation.*
 import a75f.io.renatus.R
 import android.app.Application
-import android.util.Log
 import androidx.lifecycle.AndroidViewModel
 import io.reactivex.rxjava3.subjects.BehaviorSubject
 
@@ -70,6 +68,8 @@ class HyperStatCpuViewModel(application: Application) : AndroidViewModel(applica
         if (config.analogOut2State.enabled) config.outputs.add(constructOutputAnalogOut(address, Port.ANALOG_OUT_TWO))
         if (config.analogOut3State.enabled) config.outputs.add(constructOutputAnalogOut(address, Port.ANALOG_OUT_THREE))
 
+
+
         hscProfile?.profileConfiguration?.put(address, config)
 
         if (hscConfiguration == null) {
@@ -89,7 +89,7 @@ class HyperStatCpuViewModel(application: Application) : AndroidViewModel(applica
         address: Short,
         port: Port
     ): Output {
-        var outputRelay = Output()
+        val outputRelay = Output()
         outputRelay.address = address
         outputRelay.port = port
         outputRelay.mOutputRelayActuatorType = OutputRelayActuatorType.NormallyOpen
@@ -192,18 +192,6 @@ class HyperStatCpuViewModel(application: Application) : AndroidViewModel(applica
         )
     }
 
-    fun analogOutTestSignalSelected(index: Int, position: Int) {
-        val analogOuts = currentState.analogOutUis
-        val newAnalogOuts = analogOuts.updated(
-            index, analogOuts[index].copy(testSignalPosition = position)
-        )
-        viewState.onNext(
-            currentState.copy(
-                analogOutUis = newAnalogOuts
-            )
-        )
-    }
-
 
     fun voltageAtDamperSelected(isMinPosition: Boolean, index: Int, position: Int) {
         val analogOuts = currentState.analogOutUis
@@ -310,6 +298,39 @@ class HyperStatCpuViewModel(application: Application) : AndroidViewModel(applica
             )
         )
     }
+
+
+    fun zoneVOCThresholdSelect(position: Int) {
+        viewState.onNext(
+            currentState.copy(
+                zoneVocThresholdPos = position
+            )
+        )
+    }
+    fun zoneVOCTargetSelect(position: Int) {
+        viewState.onNext(
+            currentState.copy(
+                zoneVocTargetPos = position
+            )
+        )
+    }
+
+
+    fun zonePmThresholdSelect(position: Int) {
+        viewState.onNext(
+            currentState.copy(
+                zonePm2p5ThresholdPos = position
+            )
+        )
+    }
+    fun zonePmTargetSelect(position: Int) {
+        viewState.onNext(
+            currentState.copy(
+                zonePm2p5TargetPos = position
+            )
+        )
+    }
+
 }
 
 // Dropdown choice value ranges
@@ -337,6 +358,14 @@ private val CO2Max = 2000
 
 private val INC = 10
 
+private val VOC_INC = 100
+private val PM_INC = 5
+
+private val VOCMin = 0
+private val VOCMax = 10000
+
+private val PMMin = 0
+private val PMMax = 1000
 
 private fun tempOffsetIndexFromValue(tempOffset: Double) =
     offsetIndexFromValue(TEMP_OFFSET_LIMIT_MIN, TEMP_OFFSET_INC, tempOffset)
@@ -377,39 +406,59 @@ private fun analogFanSpeedIndexFromValue(percent: Double) =
     offsetIndexFromValue(ANALOG_FAN_MIN_SPEED, ANALOG_FAN_SPEED_INC, percent)
 
 private fun analogFanSpeedFromIndex(index: Int) =
-    offsetFromIndex(ANALOG_FAN_MIN_SPEED, ANALOG_FAN_SPEED_INC.toDouble(), index)
+    offsetFromIndex(ANALOG_FAN_MIN_SPEED, ANALOG_FAN_SPEED_INC, index)
 
 fun analogFanLevelSpeedValue(): Array<String?> {
     return offsetSpinnerValues(
         ANALOG_FAN_MAX_SPEED, ANALOG_FAN_MIN_SPEED,
-        ANALOG_FAN_SPEED_INC.toDouble(), true, "%"
+        ANALOG_FAN_SPEED_INC, true, "%"
     )
 }
 
 private fun co2DCVDamperValueFromIndex(index: Int) = offsetFromIndex(CO2Min, INC.toDouble(), index)
 private fun OpeningDamperValueFromIndex(index: Int) = offsetFromIndex(OpeningRateMin, INC.toDouble(), index)
 
+private fun vocValueFromIndex(index: Int) = offsetFromIndex(VOCMin, VOC_INC.toDouble(), index)
+private fun pm25ValueFromIndex(index: Int) = offsetFromIndex(PMMin, PM_INC.toDouble(), index)
+
 
 private fun co2DCVDamperSetIndexFromValue(value: Double) = offsetIndexFromValue(CO2Min, INC.toDouble(), value)
 private fun co2DCVOpeningDamperSetIndexFromValue(value: Double) =
     offsetIndexFromValue(OpeningRateMin, INC.toDouble(), value)
 
+private fun vocSetIndexFromValue(value: Double) = offsetIndexFromValue(VOCMin, VOC_INC.toDouble(), value)
+private fun pmSetIndexFromValue(value: Double) = offsetIndexFromValue(PMMin, PM_INC.toDouble(), value)
+
 
 fun co2DCVDamperValue(): Array<String?> {
     return offsetSpinnerValues(
         CO2Max, CO2Min,
-        INC.toDouble(), true, "ppm"
+        INC.toDouble(), true, " ppm"
     )
 }
 
 fun co2DCVOpeningDamperValue(): Array<String?> {
     return offsetSpinnerValues(
         OpeningRateMax, OpeningRateMin,
-        INC.toDouble(), true, "%"
+        INC.toDouble(), true, " %"
+    )
+}
+
+fun vocValues(): Array<String?> {
+    return offsetSpinnerValues(
+        VOCMax, VOCMin,
+        VOC_INC.toDouble(), true, " ppb"
+    )
+}
+fun pmValues(): Array<String?> {
+    return offsetSpinnerValues(
+        PMMax, PMMin,
+        PM_INC.toDouble(), true, " ug/„é•"
     )
 }
 
 
+
 private fun offsetSpinnerValues(
     max: Int,
     min: Int,
@@ -426,7 +475,7 @@ private fun offsetSpinnerValues(
     for (nNum in 0 until count) {
         var rawValue = (nNum + offsetFromZeroCount).toDouble() * inc
         if (displayAsInt) {
-            nums[nNum] = String.format("%.1f", rawValue) + suffix
+            nums[nNum] = String.format("%d", rawValue.toInt()) + suffix
         } else {
             nums[nNum] = String.format("%.1f", rawValue) + suffix
         }
@@ -487,6 +536,10 @@ data class CpuViewState(
     var zoneCO2DamperOpeningRatePos: Int,
     var zoneCO2ThresholdPos: Int,
     var zoneCO2TargetPos: Int,
+    var zoneVocThresholdPos: Int,
+    var zoneVocTargetPos: Int,
+    var zonePm2p5ThresholdPos: Int,
+    var zonePm2p5TargetPos: Int,
 
     ) {
 
@@ -514,6 +567,11 @@ data class CpuViewState(
             zoneCO2DamperOpeningRatePos = co2DCVOpeningDamperSetIndexFromValue(config.zoneCO2DamperOpeningRate),
             zoneCO2ThresholdPos = co2DCVDamperSetIndexFromValue(config.zoneCO2Threshold),
             zoneCO2TargetPos = co2DCVDamperSetIndexFromValue(config.zoneCO2Target),
+            zoneVocThresholdPos =  vocSetIndexFromValue(config.zoneVOCThreshold),
+            zoneVocTargetPos =  vocSetIndexFromValue(config.zoneVOCTarget),
+            zonePm2p5ThresholdPos =  pmSetIndexFromValue(config.zonePm2p5Threshold),
+            zonePm2p5TargetPos =  pmSetIndexFromValue(config.zonePm2p5Target)
+
         )
     }
 
@@ -541,6 +599,10 @@ data class CpuViewState(
             zoneCO2DamperOpeningRate = co2DCVDamperValueFromIndex(zoneCO2DamperOpeningRatePos)
             zoneCO2Threshold = OpeningDamperValueFromIndex(zoneCO2ThresholdPos)
             zoneCO2Target = OpeningDamperValueFromIndex(zoneCO2TargetPos)
+            zoneVOCThreshold = vocValueFromIndex(zoneVocThresholdPos)
+            zoneVOCTarget = vocValueFromIndex(zoneVocTargetPos)
+            zonePm2p5Threshold = pm25ValueFromIndex(zonePm2p5ThresholdPos)
+            zonePm2p5Target = pm25ValueFromIndex(zonePm2p5TargetPos)
         }
     }
 }
diff --git a/app/src/main/res/layout/fragment_hyperstat_cpu_config.xml b/app/src/main/res/layout/fragment_hyperstat_cpu_config.xml
index de2e17e66..6b93eb567 100644
--- a/app/src/main/res/layout/fragment_hyperstat_cpu_config.xml
+++ b/app/src/main/res/layout/fragment_hyperstat_cpu_config.xml
@@ -767,6 +767,101 @@
 
 
             </LinearLayout>
+            <LinearLayout
+
+                android:paddingTop="@dimen/appbar_padding_top"
+                android:paddingBottom="@dimen/appbar_padding_top"
+                android:id="@+id/vocConfig"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_marginStart="34dp"
+                android:layout_marginTop="10dp"
+                android:layout_marginEnd="48dp">
+
+
+                <TextView
+                    android:id="@+id/zoneVocThreshold"
+                    android:layout_width="0dp"
+                    android:layout_weight="1"
+                    android:layout_height="match_parent"
+                    android:textColor="@drawable/label_textselector"
+                    android:textAppearance="@style/label_values"
+                    android:gravity="left"
+                    android:text="@string/zoneVocThreshold" />
+
+                <Spinner
+                    android:id="@+id/zoneVocThresholdSpinner"
+                    android:layout_width="0dp"
+                    android:layout_weight="1"
+                    android:layout_height="match_parent" />
+
+                <TextView
+                    android:id="@+id/zoneVocTarget"
+                    android:layout_width="0dp"
+                    android:layout_weight="1"
+                    android:layout_height="match_parent"
+                    android:textAppearance="@style/label_values"
+                    android:gravity="left"
+                    android:paddingLeft="@dimen/appbar_padding_top"
+                    android:text="@string/zoneVocTarget" />
+
+                <Spinner
+                    android:id="@+id/zoneVocTargetSpinner"
+                    android:layout_width="0dp"
+                    android:layout_weight="1"
+                    android:layout_height="match_parent" />
+
+
+            </LinearLayout>
+
+            <LinearLayout
+                android:paddingTop="@dimen/appbar_padding_top"
+                android:paddingBottom="@dimen/appbar_padding_top"
+                android:id="@+id/pmConfig"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_marginStart="34dp"
+                android:layout_marginTop="10dp"
+                android:layout_marginEnd="48dp"
+                >
+
+
+                <TextView
+                    android:id="@+id/zonepmThresholdLable"
+                    android:layout_width="0dp"
+                    android:layout_weight="1"
+                    android:layout_height="match_parent"
+                    android:textColor="@drawable/label_textselector"
+                    android:textAppearance="@style/label_values"
+                    android:gravity="left"
+                    android:text="@string/zonepmThreshold" />
+
+                <Spinner
+                    android:id="@+id/zonepmThresholdSpinner"
+                    android:layout_width="0dp"
+                    android:layout_weight="1"
+                    android:layout_height="match_parent" />
+
+                <TextView
+                    android:id="@+id/zonepmtagertLabel"
+                    android:layout_width="0dp"
+                    android:layout_weight="1"
+                    android:layout_height="match_parent"
+                    android:textAppearance="@style/label_values"
+                    android:gravity="left"
+                    android:paddingLeft="@dimen/appbar_padding_top"
+                    android:text="@string/zonepmTarget" />
+
+                <Spinner
+                    android:id="@+id/zonepmTargetSpinner"
+                    android:layout_width="0dp"
+                    android:layout_weight="1"
+                    android:layout_height="match_parent" />
+
+
+            </LinearLayout>
+
+
             <Button
                 android:id="@+id/setButton"
                 android:layout_width="wrap_content"
diff --git a/app/src/main/res/values/strings.xml b/app/src/main/res/values/strings.xml
index a9b1a389f..cd03050b0 100644
--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -136,7 +136,7 @@
     <string name="fan_enabled"> Fan Enabled </string>
     <string name="occupied_enabled"> Occupied Enabled </string>
     <string name="humidifier"> Humidifier </string>
-    <string name="dehumidifier"> De Humidifier </string>
+    <string name="dehumidifier"> DeHumidifier </string>
 
     <string-array name="hyperstat_stage_selector">
         <item>  @string/cooling_stage_1 </item>
@@ -473,12 +473,12 @@
     <string name="relay5">Relay 5</string>
     <string name="relay6">Relay 6</string>
     <string name="relay7">Relay 7</string>
-    <string name="analog1">Analog-out 1</string>
-    <string name="analog2">Analog-out 2</string>
-    <string name="analog3">Analog-out 3</string>
+    <string name="analog1">Analog-out1</string>
+    <string name="analog2">Analog-out2</string>
+    <string name="analog3">Analog-out3</string>
     <string name="door_window_sensor">Door/Window Sensor (Th2)</string>
-    <string name="analogin_1">Analog-in 1</string>
-    <string name="analogin_2">Analog-in 2</string>
+    <string name="analogin_1">Analog-in1</string>
+    <string name="analogin_2">Analog-in2</string>
 
     <string name="hyperstat_analog_out">Analog-out</string>
 
@@ -491,55 +491,55 @@
 
 
     <!--VAV Fully Modulating AHU Profile-->
-    <string name="label_analog1">Analog-out1</string>
-    <string name="label_analog2">Analog-out2</string>
-    <string name="label_analog3">Analog-out3</string>
-    <string name="label_analog4">Analog-out4</string>
+    <string name="label_analog1">Analog-Out1</string>
+    <string name="label_analog2">Analog-Out2</string>
+    <string name="label_analog3">Analog-Out3</string>
+    <string name="label_analog4">Analog-Out4</string>
     <string name="label_cooling">Cooling</string>
     <string name="label_fanspeed">Fan Speed</string>
     <string name="label_heating">Heating</string>
     <string name="label_outsideair">Outside Air</string>
     <string name="label_fanenable">Fan Enable</string>
-    <string name="label_mincooling">Analog-out1 at \nMin Cooling</string>
-    <string name="label_maxcooling">Analog-out1 at \nMax Cooling</string>
-    <string name="label_minpressure">Analog-out2 at \nMin Static Pressure</string>
-    <string name="label_maxpressure">Analog-out2 at \nMax Static Pressure</string>
-    <string name="label_minheating">Analog-out3 at \nMin Heating</string>
-    <string name="label_maxheating">Analog-out3 at \nMax Heating</string>
-    <string name="label_minfa">Analog-out4 at \nMin Fresh Air</string>
-    <string name="label_maxfa">Analog-out4 at \nMax Fresh Air</string>
+    <string name="label_mincooling">Analog-Out1 at \nMin Cooling</string>
+    <string name="label_maxcooling">Analog-Out1 at \nMax Cooling</string>
+    <string name="label_minpressure">Analog-Out2 at \nMin Static Pressure</string>
+    <string name="label_maxpressure">Analog-Out2 at \nMax Static Pressure</string>
+    <string name="label_minheating">Analog-Out3 at \nMin Heating</string>
+    <string name="label_maxheating">Analog-Out3 at \nMax Heating</string>
+    <string name="label_minfa">Analog-Out4 at \nMin Fresh Air</string>
+    <string name="label_maxfa">Analog-Out4 at \nMax Fresh Air</string>
 
 
     <!--VAV Staged RTU with VFD Fan Profile-->
-    <string name="label_economizer">Analog-out2 \nDuring Economizer</string>
-    <string name="label_recirculate">Analog-out2 \nDuring Recirculate</string>
-    <string name="label_coolstage1">Analog-out2 \nDuring Cool Stage 1</string>
-    <string name="label_coolstage2">Analog-out2 \nDuring Cool Stage 2</string>
-    <string name="label_coolstage3">Analog-out2 \nDuring Cool Stage 3</string>
-    <string name="label_coolstage4">Analog-out2 \nDuring Cool Stage 4</string>
-    <string name="label_coolstage5">Analog-out2 \nDuring Cool Stage 5</string>
-    <string name="label_heatstage1">Analog-out2 \nDuring Heat Stage 1</string>
-    <string name="label_heatstage2">Analog-out2 \nDuring Heat Stage 2</string>
-    <string name="label_heatstage3">Analog-out2 \nDuring Heat Stage 3</string>
-    <string name="label_heatstage4">Analog-out2 \nDuring Heat Stage 4</string>
-    <string name="label_heatstage5">Analog-out2 \nDuring Heat Stage 5</string>
+    <string name="label_economizer">Analog-Out2 \nDuring Economizer</string>
+    <string name="label_recirculate">Analog-Out2 \nDuring Recirculate</string>
+    <string name="label_coolstage1">Analog-Out2 \nDuring Cool Stage 1</string>
+    <string name="label_coolstage2">Analog-Out2 \nDuring Cool Stage 2</string>
+    <string name="label_coolstage3">Analog-Out2 \nDuring Cool Stage 3</string>
+    <string name="label_coolstage4">Analog-Out2 \nDuring Cool Stage 4</string>
+    <string name="label_coolstage5">Analog-Out2 \nDuring Cool Stage 5</string>
+    <string name="label_heatstage1">Analog-Out2 \nDuring Heat Stage 1</string>
+    <string name="label_heatstage2">Analog-Out2 \nDuring Heat Stage 2</string>
+    <string name="label_heatstage3">Analog-Out2 \nDuring Heat Stage 3</string>
+    <string name="label_heatstage4">Analog-Out2 \nDuring Heat Stage 4</string>
+    <string name="label_heatstage5">Analog-Out2 \nDuring Heat Stage 5</string>
 
 
-    <string name="label_vfd_analog2_min">Analog-out2 Default</string>
+    <string name="label_vfd_analog2_min">Analog-Out2 Default</string>
 
 
     <!--VAV Advanced Hybrid AHU Profile-->
     <string name="label_composite">Composite</string>
-    <string name="label_a1mincool">Analog-out1 \nAt Min Cooling</string>
-    <string name="label_a1maxcool">Analog-out1 \nAt Max Cooling</string>
-    <string name="label_a2minpres">Analog-out2 \nAt Min Static Pressure</string>
-    <string name="label_a2maxpres">Analog-out2 \nAt Max Static Pressure</string>
-    <string name="label_a3minheat">Analog-out3 \nAt Min Heating</string>
-    <string name="label_a3maxheat">Analog-out3 \nAt Max Heating</string>
-    <string name="label_a4mincool">Analog-out4 \nAt Min Cooling</string>
-    <string name="label_a4maxcool">Analog-out4 \nAt Max Cooling</string>
-    <string name="label_a4minheat">Analog-out4 \nAt Min Heating</string>
-    <string name="label_a4maxheat">Analog-out4 \nAt Max Heating</string>
+    <string name="label_a1mincool">Analog-Out1 \nAt Min Cooling</string>
+    <string name="label_a1maxcool">Analog-Out1 \nAt Max Cooling</string>
+    <string name="label_a2minpres">Analog-Out2 \nAt Min Static Pressure</string>
+    <string name="label_a2maxpres">Analog-Out2 \nAt Max Static Pressure</string>
+    <string name="label_a3minheat">Analog-Out3 \nAt Min Heating</string>
+    <string name="label_a3maxheat">Analog-Out3 \nAt Max Heating</string>
+    <string name="label_a4mincool">Analog-Out4 \nAt Min Cooling</string>
+    <string name="label_a4maxcool">Analog-Out4 \nAt Max Cooling</string>
+    <string name="label_a4minheat">Analog-Out4 \nAt Min Heating</string>
+    <string name="label_a4maxheat">Analog-Out4 \nAt Max Heating</string>
 
     <!---DCWB Profile-->
     <string name="label_dcwb_enable">DCWB Enable</string>
@@ -556,16 +556,16 @@
     <string name="label_chilled_water_max_flow_rate">Chilled water Max Flow rate (GPM) </string>
     <string name="label_analog_in1_at_valve_closed_position">Analog-In 1 at Valve Closed Position </string>
     <string name="label_analog_in1_at_valve_full_position">Analog-In 1 at Valve Full Position </string>
-    <string name="label_analog_out1_at_min_chw_valve">Analog-out 1 at Min CHW Valve </string>
-    <string name="label_analog_out1_at_max_chw_valve">Analog-out 1 at Max CHW Valve </string>
-    <string name="label_analog_out2_at_min_fan_speed">Analog-out 2 at Min Fan Speed </string>
-    <string name="label_analog_out2_at_max_fan_speed">Analog-out 2 at Max Fan Speed </string>
-    <string name="label_analog_out3_at_min_heating">Analog-out 3 at Min Heating </string>
-    <string name="label_analog_out3_at_max_heating">Analog-out 3 at Max Heating </string>
-    <string name="label_analog_out4_at_min_cooling">Analog-out 4 at Min Cooling Loop </string>
-    <string name="label_analog_out4_at_max_cooling">Analog-out 4 at Max Cooling Loop </string>
-    <string name="label_analog_out4_at_min_co2">Analog-out 4 at Min CO2 </string>
-    <string name="label_analog_out4_at_max_co2">Analog-out 4 at Max CO2 </string>
+    <string name="label_analog_out1_at_min_chw_valve">Analog-Out 1 at Min CHW Valve </string>
+    <string name="label_analog_out1_at_max_chw_valve">Analog-Out 1 at Max CHW Valve </string>
+    <string name="label_analog_out2_at_min_fan_speed">Analog-Out 2 at Min Fan Speed </string>
+    <string name="label_analog_out2_at_max_fan_speed">Analog-Out 2 at Max Fan Speed </string>
+    <string name="label_analog_out3_at_min_heating">Analog-Out 3 at Min Heating </string>
+    <string name="label_analog_out3_at_max_heating">Analog-Out 3 at Max Heating </string>
+    <string name="label_analog_out4_at_min_cooling">Analog-Out 4 at Min Cooling Loop </string>
+    <string name="label_analog_out4_at_max_cooling">Analog-Out 4 at Max Cooling Loop </string>
+    <string name="label_analog_out4_at_min_co2">Analog-Out 4 at Min CO2 </string>
+    <string name="label_analog_out4_at_max_co2">Analog-Out 4 at Max CO2 </string>
 
     <string name="label_configure_btu">Please Configure the BTU meter\'s \nModbus parameters.</string>
     <string-array name="dcwb_analog4_options">
@@ -722,8 +722,8 @@
     <string name="label_thin1">TH-in1 Input Sensor</string>
     <string name="label_useanalogin2">Use Analog-in2 for dynamic setpoint</string>
     <string name="label_analogin2">Analog-in2 Input Sensor</string>
-    <string name="label_analogout1min">Analog-out1 at Min Output</string>
-    <string name="label_analogout1max">Analog-out1 at Max Output</string>
+    <string name="label_analogout1min">Analog-Out1 at Min Output</string>
+    <string name="label_analogout1max">Analog-Out1 at Max Output</string>
     <string name="label_zeroerrormp">Expected Zero Error at Midpoint</string>
     <string name="label_targetvalue">Target Value</string>
     <string name="label_expectedError">Expected Error Range</string>
@@ -770,8 +770,8 @@
     <string name="label_co2alert">CO2 Alert Threshold</string>
     <string name="label_exfan_hysteresis">Exhaust Fan Hysteresis (%)</string>
     <string name="label_use_room_co2_sensing">Use Per Room CO2 Sensing</string>
-    <string name="label_oaoanalog1">Outside Air Damper Type\n(Analog-out 1)</string>
-    <string name="label_oaoanalog2">Return Air Damper Type\n(Analog-out 2)</string>
+    <string name="label_oaoanalog1">Outside Air Damper Type\n(Analog-Out 1)</string>
+    <string name="label_oaoanalog2">Return Air Damper Type\n(Analog-Out 2)</string>
     <string name="label_co2threshold">CO2 Threshold (ppm)</string>
     <string name="label_co2offset">CO2 Alert Offset Time</string>
     <string name="label_current_transformer_type">Current Transformer Type</string>
@@ -834,14 +834,14 @@
 
     <string name="dab_dual_duct_label_thermistor1_val">Discharge Airflow Temp</string>
 
-    <string name="dab_dual_duct_label_ao1_min_damp_heat_pos">Analog-out1 at Min\n Damper Heating Pos</string>
-    <string name="dab_dual_duct_label_ao1_max_damp_heat_pos">Analog-out1 at Max\n Damper Heating Pos</string>
-    <string name="dab_dual_duct_label_ao1_min_damp_cool_pos">Analog-out1 at Min\n Damper Cooling Pos</string>
-    <string name="dab_dual_duct_label_ao1_max_damp_cool_pos">Analog-out1 at Max\n Damper Cooling Pos</string>
-    <string name="dab_dual_duct_label_ao2_min_damp_heat_pos">Analog-out2 at Min\n Damper Heating Pos</string>
-    <string name="dab_dual_duct_label_ao2_max_damp_heat_pos">Analog-out2 at Max\n Damper Heating Pos</string>
-    <string name="dab_dual_duct_label_ao2_min_damp_cool_pos">Analog-out2 at Min\n Damper Cooling Pos</string>
-    <string name="dab_dual_duct_label_ao2_max_damp_cool_pos">Analog-out2 at Max\n Damper Cooling Pos</string>
+    <string name="dab_dual_duct_label_ao1_min_damp_heat_pos">Analog-Out1 at Min\n Damper Heating Pos</string>
+    <string name="dab_dual_duct_label_ao1_max_damp_heat_pos">Analog-Out1 at Max\n Damper Heating Pos</string>
+    <string name="dab_dual_duct_label_ao1_min_damp_cool_pos">Analog-Out1 at Min\n Damper Cooling Pos</string>
+    <string name="dab_dual_duct_label_ao1_max_damp_cool_pos">Analog-Out1 at Max\n Damper Cooling Pos</string>
+    <string name="dab_dual_duct_label_ao2_min_damp_heat_pos">Analog-Out2 at Min\n Damper Heating Pos</string>
+    <string name="dab_dual_duct_label_ao2_max_damp_heat_pos">Analog-Out2 at Max\n Damper Heating Pos</string>
+    <string name="dab_dual_duct_label_ao2_min_damp_cool_pos">Analog-Out2 at Min\n Damper Cooling Pos</string>
+    <string name="dab_dual_duct_label_ao2_max_damp_cool_pos">Analog-Out2 at Max\n Damper Cooling Pos</string>
 
     <string name="dab_dual_duct_label_co2_threshold">Zone CO2 Threshold</string>
     <string name="dab_dual_duct_label_co2_target">Zone CO2 Target</string>
@@ -1094,21 +1094,29 @@
     <string name="hello_world">Hello Square World!</string>
 
 
-    <string name="hyperstat_analog_out1_at_low_fan">Analog Out1 at\nFan Low</string>
-    <string name="hyperstat_analog_out1_at_medium_fan">Analog Out1 at\nFan Medium</string>
-    <string name="hyperstat_analog_out1_at_high_fan">Analog Out1 at\nFan High</string>
+    <string name="hyperstat_analog_out1_at_low_fan">Analog-out1 at\nFan Low</string>
+    <string name="hyperstat_analog_out1_at_medium_fan">Analog-out1 at\nFan Medium</string>
+    <string name="hyperstat_analog_out1_at_high_fan">Analog-out1 at\nFan High</string>
 
-    <string name="hyperstat_analog_out2_at_low_fan">Analog Out2 at\nFan Low</string>
-    <string name="hyperstat_analog_out2_at_medium_fan">Analog Out2 at\nFan Medium</string>
-    <string name="hyperstat_analog_out2_at_high_fan">Analog Out2 at\nFan High</string>
+    <string name="hyperstat_analog_out2_at_low_fan">Analog-out2 at\nFan Low</string>
+    <string name="hyperstat_analog_out2_at_medium_fan">Analog-out2 at\nFan Medium</string>
+    <string name="hyperstat_analog_out2_at_high_fan">Analog-out2 at\nFan High</string>
 
-    <string name="hyperstat_analog_out3_at_low_fan">Analog Out3 at\nFan Low</string>
-    <string name="hyperstat_analog_out3_at_medium_fan">Analog Out3 at\nFan Medium</string>
-    <string name="hyperstat_analog_out3_at_high_fan">Analog Out3 at\nFan High</string>
+    <string name="hyperstat_analog_out3_at_low_fan">Analog-out3 at\nFan Low</string>
+    <string name="hyperstat_analog_out3_at_medium_fan">Analog-out3 at\nFan Medium</string>
+    <string name="hyperstat_analog_out3_at_high_fan">Analog-out3 at\nFan High</string>
 
     <string name="zoneCO2DamperOpeningRate">CO2 Damper\nOpeningRate</string>
     <string name="zoneCO2Threshold">CO2 Threshold</string>
     <string name="zoneCO2Target">CO2 Target</string>
+
+    <string name="zoneVocThreshold">VOC Threshold</string>
+    <string name="zoneVocTarget">VOC Target</string>
+
+    <string name="zonepmThreshold">PM 2.5 Threshold</string>
+    <string name="zonepmTarget">PM 2.5 Target</string>
+
+
     <string-array name="hyperstate_medium_high_fanmode">
         <item>  Off </item>
         <item>  Auto </item>
diff --git a/app/src/main/res/values/styles.xml b/app/src/main/res/values/styles.xml
index cfe09b8a9..7cb19e4bc 100644
--- a/app/src/main/res/values/styles.xml
+++ b/app/src/main/res/values/styles.xml
@@ -623,7 +623,7 @@
     </style>
 
     <style name="DamperPositionLabel_HyperStat">
-        <item name="android:layout_width">wrap_content</item>
+        <item name="android:layout_width">250dp</item>
         <item name="android:layout_height">wrap_content</item>
         <item name="android:layout_marginBottom">20dp</item>
         <item name="android:layout_marginTop">14dp</item>
@@ -671,4 +671,25 @@
         <item name="android:layout_gravity">end|top</item>
         <item name="android:dropDownWidth">match_parent</item>
     </style>
+
+    <style name="VocLabel_HyperStat">
+        <item name="android:layout_width">0dp</item>
+        <item name="android:layout_weight">1</item>
+        <item name="android:layout_height">wrap_content</item>
+        <item name="android:layout_marginBottom">20dp</item>
+        <item name="android:layout_marginTop">14dp</item>
+        <item name="android:textAppearance">@style/label_values</item>
+        <item name="android:textColor">@drawable/label_textselector</item>
+    </style>
+    <style name="VocHyperStat" parent="Widget.AppCompat.Spinner">
+        <item name="android:layout_width">0dp</item>
+        <item name="android:layout_weight">1.5</item>
+        <item name="android:layout_marginBottom">20dp</item>
+        <item name="android:layout_marginTop">20dp</item>
+        <item name="android:layout_marginRight">10dp</item>
+        <item name="android:layout_marginStart">10dp</item>
+        <item name="android:layout_height">wrap_content</item>
+        <item name="android:layout_gravity">end|top</item>
+        <item name="android:dropDownWidth">match_parent</item>
+    </style>
 </resources>
diff --git a/device/src/main/java/a75f/io/device/mesh/hyperstat/HyperStatMessageGenerator.java b/device/src/main/java/a75f/io/device/mesh/hyperstat/HyperStatMessageGenerator.java
index 9857a4100..319f2001c 100644
--- a/device/src/main/java/a75f/io/device/mesh/hyperstat/HyperStatMessageGenerator.java
+++ b/device/src/main/java/a75f/io/device/mesh/hyperstat/HyperStatMessageGenerator.java
@@ -77,8 +77,7 @@ public class HyperStatMessageGenerator {
      * @return
      */
     public static HyperStatSettingsMessage_t getSettingsMessage(String zone, int address, String equipRef) {
-        //TODO - Proto file does not define profile bitmap, enabledRelay.
-        HyperStatSettingsMessage_t settings = HyperStatSettingsMessage_t.newBuilder()
+        return HyperStatSettingsMessage_t.newBuilder()
             .setRoomName(zone)
             .setHeatingDeadBand((int) (getStandaloneHeatingDeadband(equipRef) * 10))
             .setCoolingDeadBand((int) (getStandaloneCoolingDeadband(equipRef) * 10))
@@ -89,11 +88,13 @@ public class HyperStatMessageGenerator {
             .setTemperatureOffset((int) (DeviceHSUtil.getTempOffset(address)))
             .setHumidityMinSetpoint(getHumidityMinSp(address, CCUHsApi.getInstance()))
             .setHumidityMaxSetpoint(getHumidityMaxSp(address, CCUHsApi.getInstance()))
-             .setDisplayHumidity(true)
-             .setDisplayCO2(true)
+            .setDisplayHumidity(true)
+            .setDisplayCO2(true)
+            .setCo2AlertThreshold((int)readCo2ThresholdValue(equipRef))
+            .setPm25AlertThreshold((int)readVocThresholdValue(equipRef))
+            .setVocAlertThreshold((int)readPm2p5ThresholdValue(equipRef))
             .setTemperatureMode(HyperStat.HyperStatTemperatureMode_e.HYPERSTAT_TEMP_MODE_DUAL_VARIABLE_DB)
             .build();
-        return settings;
     }
     
     /**
@@ -295,6 +296,22 @@ public class HyperStatMessageGenerator {
         }
 
     }
+    public static double readCo2ThresholdValue(String equipRef) {
+        return CCUHsApi.getInstance().readDefaultVal(
+                "point and hyperstat and co2 and threshold and equipRef == \""+equipRef+ "\"");
+    }
+    public static double readVocThresholdValue(String equipRef) {
+        return CCUHsApi.getInstance().readDefaultVal(
+                "point and hyperstat and voc and threshold and equipRef == \""+equipRef+ "\"");
+    }
+
+    public static double readPm2p5ThresholdValue(String equipRef) {
+        return CCUHsApi.getInstance().readDefaultVal(
+                "point and hyperstat and pm2p5 and threshold and equipRef == \""+equipRef+ "\"");
+    }
+
+
+
 
     public static HyperStat.HyperStatSettingsMessage2_t getSetting2Message(int address, String equipRef){
         return  HyperStatSettingsUtil.Companion.getSetting2Message(address,equipRef,CCUHsApi.getInstance());
diff --git a/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/common/HSReconfigureUtil.kt b/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/common/HSReconfigureUtil.kt
index abf4bad88..14e11ee4b 100644
--- a/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/common/HSReconfigureUtil.kt
+++ b/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/common/HSReconfigureUtil.kt
@@ -135,7 +135,7 @@ class HSReconfigureUtil {
         }
 
         // Get Equip Points Util
-        fun getEquipPointsUtil(equip: Equip, haystack: CCUHsApi): HyperStatPointsUtil {
+         fun getEquipPointsUtil(equip: Equip, haystack: CCUHsApi): HyperStatPointsUtil {
             return HyperStatPointsUtil(
                 profileName = CPU,
                 equipRef = equip.id,
diff --git a/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/cpu/HyperStatCpuConfiguration.kt b/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/cpu/HyperStatCpuConfiguration.kt
index a18ecefba..9c5609188 100644
--- a/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/cpu/HyperStatCpuConfiguration.kt
+++ b/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/cpu/HyperStatCpuConfiguration.kt
@@ -36,6 +36,11 @@ class HyperStatCpuConfiguration : BaseProfileConfiguration() {
    var zoneCO2Threshold = 800.0
    var zoneCO2Target = 1000.0
 
+   var zoneVOCThreshold = 10000.0
+   var zoneVOCTarget = 10000.0
+   var zonePm2p5Threshold = 1000.0
+   var zonePm2p5Target = 1000.0
+
    companion object {
       fun default(): HyperStatCpuConfiguration = HyperStatCpuConfiguration()
    }
diff --git a/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/cpu/HyperStatCpuEquip.kt b/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/cpu/HyperStatCpuEquip.kt
index 61054a8bd..39bbb67fe 100644
--- a/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/cpu/HyperStatCpuEquip.kt
+++ b/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/cpu/HyperStatCpuEquip.kt
@@ -287,6 +287,16 @@ class HyperStatCpuEquip(val node: Short) {
             .createPointCO2ConfigPoint(
                 hyperStatConfig = hyperStatConfig
             )
+        val vocPmPointsList: MutableList<Pair<Point, Any>> = hyperStatPointsUtil
+            .createPointVOCPmConfigPoint(
+                equipDis = equipDis!!,
+                zonePm2p5Target = hyperStatConfig.zonePm2p5Target,
+                zonePm2p5Threshold = hyperStatConfig.zonePm2p5Threshold,
+                zoneVOCTarget = hyperStatConfig.zoneVOCTarget,
+                zoneVOCThreshold = hyperStatConfig.zoneVOCThreshold
+            )
+
+
         val loopOutputPoints: MutableList<Pair<Point, Any>> = hyperStatPointsUtil.createLoopOutputPoints()
 
         val relayConfigPoints: MutableList<Pair<Point, Any>> = hyperStatPointsUtil.createIsRelayEnabledConfigPoints(
@@ -311,7 +321,8 @@ class HyperStatCpuEquip(val node: Short) {
 
         val allConfigPoints = arrayOf(
             configPointsList, relayConfigPoints, analogOutConfigPoints,
-            analogInConfigPoints, thConfigPointsList, userIntentPointsList, co2ConfigPointsList, loopOutputPoints
+            analogInConfigPoints, thConfigPointsList, userIntentPointsList,
+            co2ConfigPointsList, loopOutputPoints,vocPmPointsList
         )
         Log.i(L.TAG_CCU_HSCPU, "adding : points default value ")
         hyperStatPointsUtil.addPointsListToHaystackWithDefaultValue(listOfAllPoints = allConfigPoints)
@@ -361,114 +372,166 @@ class HyperStatCpuEquip(val node: Short) {
     fun updateConfigPoints(updatedHyperStatConfig: HyperStatCpuConfiguration) {
 
         val presetConfiguration = getConfiguration()
+        Log.i(L.TAG_CCU_HSCPU, "===========Hyperstat profile Update  ============")
+        updateGeneralConfiguration(newConfiguration = updatedHyperStatConfig, existingConfiguration = presetConfiguration)
+        updateAutoAwayConfiguration(newConfiguration = updatedHyperStatConfig, existingConfiguration = presetConfiguration)
+        updateAutoAForceOccupyConfiguration(newConfiguration = updatedHyperStatConfig, existingConfiguration = presetConfiguration)
+        updateAirFlowTempSensorConfiguration(newConfiguration = updatedHyperStatConfig, existingConfiguration = presetConfiguration)
+        updateDoorWindowSensorConfiguration(newConfiguration = updatedHyperStatConfig, existingConfiguration = presetConfiguration)
+        updateRelaysConfig(newConfiguration = updatedHyperStatConfig, existingConfiguration = presetConfiguration)
+        updateAnalogOutConfig(newConfiguration = updatedHyperStatConfig, existingConfiguration = presetConfiguration)
+        updateAnalogInConfig(newConfiguration = updatedHyperStatConfig, existingConfiguration = presetConfiguration)
+        Log.i(L.TAG_CCU_HSCPU, "Profile update has been completed  ")
+        haystack.syncEntityTree()
 
-        if (updatedHyperStatConfig.temperatureOffset != presetConfiguration.temperatureOffset) {
-            val tempOffId = hsHaystackUtil!!.readPointID("temperature and offset") as String
-            hyperStatPointsUtil.addDefaultValueForPoint(tempOffId, (updatedHyperStatConfig.temperatureOffset * 10))
-        }
-        if (updatedHyperStatConfig.zoneCO2DamperOpeningRate != presetConfiguration.zoneCO2DamperOpeningRate) {
-            val pointId = hsHaystackUtil!!.readPointID("co2 and opening and rate") as String
-            hyperStatPointsUtil.addDefaultValueForPoint(pointId, updatedHyperStatConfig.zoneCO2DamperOpeningRate)
-            hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, updatedHyperStatConfig.zoneCO2DamperOpeningRate)
-        }
-        if (updatedHyperStatConfig.zoneCO2Threshold != presetConfiguration.zoneCO2Threshold) {
-            val pointId = hsHaystackUtil!!.readPointID("co2 and threshold") as String
-            hyperStatPointsUtil.addDefaultValueForPoint(pointId, updatedHyperStatConfig.zoneCO2Threshold)
-            hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, updatedHyperStatConfig.zoneCO2Threshold)
-        }
-        if (updatedHyperStatConfig.zoneCO2Target != presetConfiguration.zoneCO2Target) {
-            val pointId = hsHaystackUtil!!.readPointID("co2 and target") as String
-            hyperStatPointsUtil.addDefaultValueForPoint(pointId, updatedHyperStatConfig.zoneCO2Target)
-            hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, updatedHyperStatConfig.zoneCO2Target)
-        }
-
-
-        if (updatedHyperStatConfig.isEnableAutoForceOccupied != presetConfiguration.isEnableAutoForceOccupied) {
-            val autoForceOccupiedEnabledId = hsHaystackUtil!!.readPointID(
-                "config and auto and occupancy and forced and control"
-            ) as String
-            hyperStatPointsUtil.addDefaultValueForPoint(
-                autoForceOccupiedEnabledId, if (updatedHyperStatConfig.isEnableAutoForceOccupied) 1.0 else 0.0
-            )
-            hyperStatPointsUtil.addDefaultHisValueForPoint(
-                autoForceOccupiedEnabledId,if (updatedHyperStatConfig.isEnableAutoForceOccupied) 1.0 else 0.0
-            )
-        }
-
-        if (updatedHyperStatConfig.isEnableAutoAway != presetConfiguration.isEnableAutoAway) {
-            val autoAwayEnabledId = hsHaystackUtil!!.readPointID(
-                "config and  auto and away"
-            ) as String
-            hyperStatPointsUtil.addDefaultValueForPoint(
-                autoAwayEnabledId, if (updatedHyperStatConfig.isEnableAutoAway) 1.0 else 0.0
-            )
-            hyperStatPointsUtil.addDefaultHisValueForPoint(
-                autoAwayEnabledId,if (updatedHyperStatConfig.isEnableAutoAway) 1.0 else 0.0
-            )
-            hsHaystackUtil!!.reWriteOccupancy(equipRef!!)
-        }
-        if (updatedHyperStatConfig.isEnableAirFlowTempSensor != presetConfiguration.isEnableAirFlowTempSensor) {
+    }
+    // Function to update DoorWindowSensor configuration
+    private fun updateDoorWindowSensorConfiguration(
+        newConfiguration: HyperStatCpuConfiguration,
+        existingConfiguration: HyperStatCpuConfiguration){
+        if (newConfiguration.isEnableDoorWindowSensor != existingConfiguration.isEnableDoorWindowSensor) {
             val enabledId = hsHaystackUtil!!.readPointID(
-                "config and airflow and temp and th1"
+                "config and window and sensor and th2"
             ) as String
             hyperStatPointsUtil.addDefaultValueForPoint(
-                enabledId, if (updatedHyperStatConfig.isEnableAirFlowTempSensor) 1.0 else 0.0
+                enabledId, if (newConfiguration.isEnableDoorWindowSensor) 1.0 else 0.0
             )
             val logicalPointId: String? = hsHaystackUtil!!.readPointID(
-                "th1 and airflow and discharge and logical"
+                "th2 and window and sensor and logical"
             )
             if (logicalPointId != null) {
                 hsHaystackUtil!!.removePoint(logicalPointId)
             }
-            if (updatedHyperStatConfig.isEnableAirFlowTempSensor) {
-                val pointData: Point = hyperStatPointsUtil.createPointForAirflowTempSensor()
+            if (newConfiguration.isEnableDoorWindowSensor) {
+                val pointData: Point = hyperStatPointsUtil.createPointForDoorWindowSensor("th2")
                 val pointId = hyperStatPointsUtil.addPointToHaystack(pointData)
                 if (pointData.markers.contains("his")) {
                     hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, 0.0)
                 }
                 hyperStatPointsUtil.addDefaultValueForPoint(pointId, 0.0)
-                DeviceUtil.setPointEnabled(nodeAddress.toInt(), Port.TH1_IN.name, true)
-                DeviceUtil.updatePhysicalPointRef(nodeAddress.toInt(), Port.TH1_IN.name, pointId)
+                DeviceUtil.setPointEnabled(nodeAddress.toInt(), Port.TH2_IN.name, true)
+                DeviceUtil.updatePhysicalPointRef(nodeAddress.toInt(), Port.TH2_IN.name, pointId)
+                DeviceUtil.updatePhysicalPointType(
+                    nodeAddress.toInt(), Port.TH2_IN.name,
+                    HyperStatAssociationUtil.getSensorNameByType(CpuAnalogInAssociation.DOOR_WINDOW_SENSOR)
+                )
             }
-
         }
-        if (updatedHyperStatConfig.isEnableDoorWindowSensor != presetConfiguration.isEnableDoorWindowSensor) {
+    }
+
+    // Function to update AirFlowTempSensor configuration
+    private fun updateAirFlowTempSensorConfiguration(
+        newConfiguration: HyperStatCpuConfiguration,
+        existingConfiguration: HyperStatCpuConfiguration) {
+        if (newConfiguration.isEnableAirFlowTempSensor != existingConfiguration.isEnableAirFlowTempSensor) {
             val enabledId = hsHaystackUtil!!.readPointID(
-                "config and window and sensor and th2"
+                "config and airflow and temp and th1"
             ) as String
             hyperStatPointsUtil.addDefaultValueForPoint(
-                enabledId, if (updatedHyperStatConfig.isEnableDoorWindowSensor) 1.0 else 0.0
+                enabledId, if (newConfiguration.isEnableAirFlowTempSensor) 1.0 else 0.0
             )
             val logicalPointId: String? = hsHaystackUtil!!.readPointID(
-                "th2 and window and sensor and logical"
+                "th1 and airflow and discharge and logical"
             )
             if (logicalPointId != null) {
                 hsHaystackUtil!!.removePoint(logicalPointId)
             }
-            if (updatedHyperStatConfig.isEnableDoorWindowSensor) {
-                val pointData: Point = hyperStatPointsUtil.createPointForDoorWindowSensor("th2")
+            if (newConfiguration.isEnableAirFlowTempSensor) {
+                val pointData: Point = hyperStatPointsUtil.createPointForAirflowTempSensor()
                 val pointId = hyperStatPointsUtil.addPointToHaystack(pointData)
                 if (pointData.markers.contains("his")) {
                     hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, 0.0)
                 }
                 hyperStatPointsUtil.addDefaultValueForPoint(pointId, 0.0)
-                DeviceUtil.setPointEnabled(nodeAddress.toInt(), Port.TH2_IN.name, true)
-                DeviceUtil.updatePhysicalPointRef(nodeAddress.toInt(), Port.TH2_IN.name, pointId)
-                DeviceUtil.updatePhysicalPointType(
-                    nodeAddress.toInt(), Port.TH2_IN.name,
-                    HyperStatAssociationUtil.getSensorNameByType(CpuAnalogInAssociation.DOOR_WINDOW_SENSOR)
-                )
+                DeviceUtil.setPointEnabled(nodeAddress.toInt(), Port.TH1_IN.name, true)
+                DeviceUtil.updatePhysicalPointRef(nodeAddress.toInt(), Port.TH1_IN.name, pointId)
             }
         }
-        Log.i(L.TAG_CCU_HSCPU, "===========Hyperstat profile Update  ============")
-        updateRelaysConfig(newConfiguration = updatedHyperStatConfig, existingConfiguration = presetConfiguration)
-        updateAnalogOutConfig(newConfiguration = updatedHyperStatConfig, existingConfiguration = presetConfiguration)
-        updateAnalogInConfig(newConfiguration = updatedHyperStatConfig, existingConfiguration = presetConfiguration)
-        Log.i(L.TAG_CCU_HSCPU, "Profile update has been completed  ")
-        haystack.syncEntityTree()
+    }
+
+    // Function to update Auto Force Occupy configuration
+    private fun updateAutoAForceOccupyConfiguration(
+        newConfiguration: HyperStatCpuConfiguration,
+        existingConfiguration: HyperStatCpuConfiguration) {
+        if (newConfiguration.isEnableAutoForceOccupied != existingConfiguration.isEnableAutoForceOccupied) {
+            val autoForceOccupiedEnabledId = hsHaystackUtil!!.readPointID(
+                "config and auto and occupancy and forced and control"
+            ) as String
+            hyperStatPointsUtil.addDefaultValueForPoint(
+                autoForceOccupiedEnabledId, if (newConfiguration.isEnableAutoForceOccupied) 1.0 else 0.0
+            )
+            hyperStatPointsUtil.addDefaultHisValueForPoint(
+                autoForceOccupiedEnabledId,if (newConfiguration.isEnableAutoForceOccupied) 1.0 else 0.0
+            )
+        }
 
     }
 
+    // Function to update AutoAway configuration
+    private fun updateAutoAwayConfiguration(
+    newConfiguration: HyperStatCpuConfiguration,
+    existingConfiguration: HyperStatCpuConfiguration) {
+        if (newConfiguration.isEnableAutoAway != existingConfiguration.isEnableAutoAway) {
+            val autoAwayEnabledId = hsHaystackUtil!!.readPointID(
+                "config and  auto and away"
+            ) as String
+            hyperStatPointsUtil.addDefaultValueForPoint(
+                autoAwayEnabledId, if (newConfiguration.isEnableAutoAway) 1.0 else 0.0
+            )
+            hyperStatPointsUtil.addDefaultHisValueForPoint(
+                autoAwayEnabledId,if (newConfiguration.isEnableAutoAway) 1.0 else 0.0
+            )
+            hsHaystackUtil!!.reWriteOccupancy(equipRef!!)
+        }
+    }
+
+    // function to update (If change in configuration)
+    // Temp Off set , CO2 threshold and target, VOC threshold and target, pm2.5 threshold and target
+    private fun updateGeneralConfiguration(
+        newConfiguration: HyperStatCpuConfiguration,
+        existingConfiguration: HyperStatCpuConfiguration){
+
+        if (newConfiguration.temperatureOffset != existingConfiguration.temperatureOffset) {
+            val tempOffId = hsHaystackUtil!!.readPointID("temperature and offset") as String
+            hyperStatPointsUtil.addDefaultValueForPoint(tempOffId, (newConfiguration.temperatureOffset * 10))
+        }
+        if (newConfiguration.zoneCO2DamperOpeningRate != existingConfiguration.zoneCO2DamperOpeningRate) {
+            val pointId = hsHaystackUtil!!.readPointID("co2 and opening and rate") as String
+            hyperStatPointsUtil.addDefaultValueForPoint(pointId, newConfiguration.zoneCO2DamperOpeningRate)
+            hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, newConfiguration.zoneCO2DamperOpeningRate)
+        }
+        if (newConfiguration.zoneCO2Threshold != existingConfiguration.zoneCO2Threshold) {
+            val pointId = hsHaystackUtil!!.readPointID("co2 and threshold") as String
+            hyperStatPointsUtil.addDefaultValueForPoint(pointId, newConfiguration.zoneCO2Threshold)
+            hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, newConfiguration.zoneCO2Threshold)
+        }
+        if (newConfiguration.zoneCO2Target != existingConfiguration.zoneCO2Target) {
+            val pointId = hsHaystackUtil!!.readPointID("co2 and target") as String
+            hyperStatPointsUtil.addDefaultValueForPoint(pointId, newConfiguration.zoneCO2Target)
+            hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, newConfiguration.zoneCO2Target)
+        }
+        if (newConfiguration.zoneVOCThreshold != existingConfiguration.zoneVOCThreshold) {
+            val pointId = hsHaystackUtil!!.readPointID("voc and threshold") as String
+            hyperStatPointsUtil.addDefaultValueForPoint(pointId, newConfiguration.zoneVOCThreshold)
+            hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, newConfiguration.zoneVOCThreshold)
+        }
+        if (newConfiguration.zoneVOCTarget != existingConfiguration.zoneVOCTarget) {
+            val pointId = hsHaystackUtil!!.readPointID("voc and target") as String
+            hyperStatPointsUtil.addDefaultValueForPoint(pointId, newConfiguration.zoneVOCTarget)
+            hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, newConfiguration.zoneVOCTarget)
+        }
+        if (newConfiguration.zonePm2p5Threshold != existingConfiguration.zonePm2p5Threshold) {
+            val pointId = hsHaystackUtil!!.readPointID("pm2p5 and threshold") as String
+            hyperStatPointsUtil.addDefaultValueForPoint(pointId, newConfiguration.zonePm2p5Threshold)
+            hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, newConfiguration.zonePm2p5Threshold)
+        }
+        if (newConfiguration.zonePm2p5Target != existingConfiguration.zonePm2p5Target) {
+            val pointId = hsHaystackUtil!!.readPointID("pm2p5 and target") as String
+            hyperStatPointsUtil.addDefaultValueForPoint(pointId, newConfiguration.zonePm2p5Target)
+            hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, newConfiguration.zonePm2p5Target)
+        }
+
+    }
 
     // collect the existing profile configurations
     fun getConfiguration(): HyperStatCpuConfiguration {
@@ -512,6 +575,22 @@ class HyperStatCpuEquip(val node: Short) {
             "point and hyperstat and co2 and target and equipRef == \"$equipRef\""
         )
 
+        config.zoneVOCThreshold = haystack.readDefaultVal(
+            "point and hyperstat and voc and threshold and air and equipRef == \"$equipRef\""
+        )
+
+        config.zoneVOCTarget = haystack.readDefaultVal(
+            "point and hyperstat and voc and target and air and equipRef == \"$equipRef\""
+        )
+
+        config.zonePm2p5Threshold = haystack.readDefaultVal(
+            "point and hyperstat and pm2p5 and threshold and air and equipRef == \"$equipRef\""
+        )
+
+        config.zonePm2p5Target = haystack.readDefaultVal(
+            "point and hyperstat and pm2p5 and target and air and equipRef == \"$equipRef\""
+        )
+
         return config
     }
 
diff --git a/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/cpu/HyperStatPointsUtil.kt b/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/cpu/HyperStatPointsUtil.kt
index d8ab7b220..eab701588 100644
--- a/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/cpu/HyperStatPointsUtil.kt
+++ b/logic/src/main/java/a75f/io/logic/bo/building/hyperstat/cpu/HyperStatPointsUtil.kt
@@ -796,7 +796,64 @@ class HyperStatPointsUtil constructor(
     }
 
 
-    // Function which creates User Intent Points
+    // Function to create VOV and pm2.5 sensor points
+    fun createPointVOCPmConfigPoint(
+        equipDis: String,
+        zoneVOCThreshold: Double,
+        zoneVOCTarget: Double,
+        zonePm2p5Threshold: Double,
+        zonePm2p5Target: Double
+    ): MutableList<Pair<Point, Any>> {
+        val pointsList: MutableList<Pair<Point, Any>> = LinkedList()
+        //ppb
+        val vocMarkers: MutableList<String> = LinkedList()
+        val pm2p5Markers: MutableList<String> = LinkedList()
+
+        vocMarkers.addAll(arrayOf(
+            "zone","cov","air","voc","sensor","current","his","cur","logical","config", "writable", "his","threshold"))
+        pm2p5Markers.addAll(arrayOf(
+            "zone","cov","air","pm2p5","sensor","current","his","cur","logical","config", "writable", "his","threshold"))
+
+
+        val zoneVOCThresholdPoint = createHaystackPointWithUnit(
+            "$equipDis-zoneVOCThreshold",
+            vocMarkers.stream().toArray { arrayOfNulls(it) },
+            "","ppb"
+        )
+        vocMarkers.remove("threshold")
+        vocMarkers.add("target")
+        val zoneVOCTargetPoint = createHaystackPointWithUnit(
+            "$equipDis-zoneVOCTarget",
+            vocMarkers.stream().toArray { arrayOfNulls(it) },
+            "","ppb"
+        )
+
+        val zonePm2p5ThresholdPoint = createHaystackPointWithUnit(
+            "$equipDis-zonePm2p5Threshold",
+            pm2p5Markers.stream().toArray { arrayOfNulls(it) },
+            "","ug/\u33A5"
+        )
+
+        pm2p5Markers.remove("threshold")
+        pm2p5Markers.add("target")
+        val zonePm2p5TargetPoint = createHaystackPointWithUnit(
+            "$equipDis-zonePm2p5Target",
+            pm2p5Markers.stream().toArray { arrayOfNulls(it) },
+            "","ug/\u33A5"
+        )
+
+
+        pointsList.add(Pair(zoneVOCThresholdPoint, zoneVOCThreshold))
+        pointsList.add(Pair(zoneVOCTargetPoint, zoneVOCTarget))
+
+        pointsList.add(Pair(zonePm2p5ThresholdPoint, zonePm2p5Threshold))
+        pointsList.add(Pair(zonePm2p5TargetPoint, zonePm2p5Target))
+
+        return pointsList
+    }
+
+
+        // Function which creates User Intent Points
     fun createUserIntentPoints(hyperStatConfig: HyperStatCpuConfiguration): MutableList<Pair<Point, Any>> {
 
         val userIntentPointList: MutableList<Pair<Point, Any>> = LinkedList()
@@ -1697,6 +1754,7 @@ class HyperStatPointsUtil constructor(
         return 0
     }
 
+
 }
 
 
diff --git a/logic/src/main/java/a75f/io/logic/util/MigrationUtil.java b/logic/src/main/java/a75f/io/logic/util/MigrationUtil.java
index 6ca4a9495..d86279772 100644
--- a/logic/src/main/java/a75f/io/logic/util/MigrationUtil.java
+++ b/logic/src/main/java/a75f/io/logic/util/MigrationUtil.java
@@ -33,9 +33,12 @@ import a75f.io.logic.bo.building.definitions.ProfileType;
 import a75f.io.logic.bo.building.definitions.ScheduleType;
 import a75f.io.logic.bo.building.definitions.Units;
 import a75f.io.logic.bo.building.dualduct.DualDuctEquip;
+import a75f.io.logic.bo.building.hyperstat.common.HSReconfigureUtil;
+import a75f.io.logic.bo.building.hyperstat.cpu.HyperStatPointsUtil;
 import a75f.io.logic.bo.building.vav.VavEquip;
 import a75f.io.logic.bo.haystack.device.SmartNode;
 import a75f.io.logic.diag.DiagEquip;
+import kotlin.Pair;
 
 public class MigrationUtil {
     
@@ -84,9 +87,8 @@ public class MigrationUtil {
             airflowUnitMigration(CCUHsApi.getInstance());
             PreferenceUtil.setAirflowVolumeUnitMigrationDone();
         }
-        Log.i(TAG_CCU_MIGRATION_UTIL, "doDamperFeedbackMigration: check "+PreferenceUtil.getDamperFeedbackMigration());
+
         if (!PreferenceUtil.getDamperFeedbackMigration()) {
-            Log.i(TAG_CCU_MIGRATION_UTIL, "doDamperFeedbackMigration: check ");
             doDamperFeedbackMigration(CCUHsApi.getInstance());
              PreferenceUtil.setDamperFeedbackMigration();
         }
@@ -95,6 +97,54 @@ public class MigrationUtil {
             addUnitToTuners(CCUHsApi.getInstance());
             PreferenceUtil.setUnitAddedToTuners();
         }
+
+        if(!PreferenceUtil.getVocPm2p5Migration()){
+            migrateVocPm2p5(CCUHsApi.getInstance());
+            PreferenceUtil.setVocPm2p5Migration();
+        }
+
+
+
+    }
+
+    private static void migrateVocPm2p5(CCUHsApi instance) {
+        ArrayList<HashMap<Object, Object>> hyperstatEquips = instance.readAllEntities("equip and hyperstat");
+        hyperstatEquips.forEach(rawEquip -> {
+            Equip equip = new Equip.Builder().setHashMap(rawEquip).build();
+
+            boolean isCovThresholdExist = isPointExist ("point and hyperstat and cov and threshold and equipRef == \"" +equip.getId()+"\"" ,instance);
+            boolean isCovTargetExist = isPointExist ("point and hyperstat and cov and target and equipRef == \"" +equip.getId()+"\"" ,instance);
+            boolean isPm2p5ThresholdExist = isPointExist ("point and hyperstat and pm2p5 and threshold and equipRef == \"" +equip.getId()+"\"" ,instance);
+            boolean isPm2p5TargetExist = isPointExist ("point and hyperstat and pm2p5 and target and equipRef == \"" +equip.getId()+"\"" ,instance);
+
+            HyperStatPointsUtil hyperStatPointsUtil = HSReconfigureUtil.Companion.getEquipPointsUtil(equip, instance);
+
+            List<Pair<Point, Object>> list = hyperStatPointsUtil.createPointVOCPmConfigPoint(
+                    equip.getDisplayName(), 1000, 1000, 1000, 1000
+            );
+            list.forEach(pointObjectPair -> {
+                if((pointObjectPair.getFirst().getDisplayName().contains("zoneVOCThreshold") && !isCovThresholdExist)
+                        ||(pointObjectPair.getFirst().getDisplayName().contains("zoneVOCTarget") && !isCovTargetExist)
+                        ||(pointObjectPair.getFirst().getDisplayName().contains("zonePm2p5Threshold") && !isPm2p5ThresholdExist)
+                        ||(pointObjectPair.getFirst().getDisplayName().contains("zonePm2p5Target") && !isPm2p5TargetExist)){
+                    pushPointToHS(hyperStatPointsUtil,pointObjectPair);
+                }
+            });
+        });
+    }
+
+    private static void pushPointToHS(HyperStatPointsUtil hyperStatPointsUtil, Pair<Point, Object> pointDetails ){
+        String pointId = hyperStatPointsUtil.addPointToHaystack(pointDetails.getFirst());
+        if (pointDetails.getFirst().getMarkers().contains("his")) {
+            hyperStatPointsUtil.addDefaultHisValueForPoint(pointId, pointDetails.getSecond());
+        }
+        if (pointDetails.getFirst().getMarkers().contains("writable")) {
+            hyperStatPointsUtil.addDefaultValueForPoint(pointId, pointDetails.getSecond());
+        }
+    }
+
+    private static boolean isPointExist(String query, CCUHsApi ccuHsApi){
+      return !ccuHsApi.readEntity(query).isEmpty();
     }
 
     private static void airflowUnitMigration(CCUHsApi ccuHsApi) {
diff --git a/logic/src/main/java/a75f/io/logic/util/PreferenceUtil.java b/logic/src/main/java/a75f/io/logic/util/PreferenceUtil.java
index 674a84f01..3027789f4 100644
--- a/logic/src/main/java/a75f/io/logic/util/PreferenceUtil.java
+++ b/logic/src/main/java/a75f/io/logic/util/PreferenceUtil.java
@@ -19,6 +19,7 @@ public class PreferenceUtil {
     private static final String ENABLE_ZONE_SCHEDULE_MIGRATION = "enableZoneScheduleMigration";
     private static final String CLEAN_UP_DUPLICATE_ZONE_SCHEDULE = "cleanUpDuplicateZoneSchedule";
     private static final String DAMPER_FEEDBACK_MIGRATION = "damperFeedbackMigration";
+    private static final String VOC_PM2P5_MIGRATION = "VovPm2p5Migration";
 
     public static void setContext(Context c) {
         context= c;
@@ -219,16 +220,25 @@ public class PreferenceUtil {
     public static boolean getSNMigration() {
         return getBooleanPreference(SMART_NODE_MIGRATION);
     }
+
+    public static void setSmartNodeMigration() {
+        setBooleanPreference(SMART_NODE_MIGRATION, true);
+
+    }
+
     public static void setDamperFeedbackMigration() {
-        setBooleanPreference(DAMPER_FEEDBACK_MIGRATION, false);
+        setBooleanPreference(DAMPER_FEEDBACK_MIGRATION, true);
     }
     public static boolean getDamperFeedbackMigration() {
         return getBooleanPreference(DAMPER_FEEDBACK_MIGRATION);
     }
+    public static void setVocPm2p5Migration() {
+        setBooleanPreference(VOC_PM2P5_MIGRATION, true);
+    }
+    public static boolean getVocPm2p5Migration() {
+        return getBooleanPreference(VOC_PM2P5_MIGRATION);
+    }
 
 
-    public static void setSmartNodeMigration() {
-        setBooleanPreference(SMART_NODE_MIGRATION, true);
 
-    }
 }
\ No newline at end of file
-- 
2.31.0.windows.1

